### ISSUE TITLE
exec: a crash with vectorized distributed

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#38918]
Found when running `fakedist` config on `cascade` logic test with "streaming" by default.

To reproduce, I created a 3 node local cluster with (in 4 separate tabs):
`./cockroach start --insecure --store=node1 --listen-addr=localhost:26257 --http-addr=localhost:8080 --join=localhost:26257,localhost:26258,localhost:26259`
`./cockroach start --insecure --store=node2 --listen-addr=localhost:26258 --http-addr=localhost:8081 --join=localhost:26257,localhost:26258,localhost:26259`
`./cockroach start --insecure --store=node3 --listen-addr=localhost:26259 --http-addr=localhost:8082 --join=localhost:26257,localhost:26258,localhost:26259`
`./cockroach init --insecure --host=localhost:26257`

And run the following commands:
```
SET experimental_vectorize=on;
CREATE TABLE a (
  id INT PRIMARY KEY
);
CREATE TABLE b (
  id INT PRIMARY KEY REFERENCES a ON UPDATE CASCADE
 ,CONSTRAINT less_than_1000 CHECK (id < 1000)
);
CREATE TABLE c (
  id INT PRIMARY KEY REFERENCES b ON UPDATE CASCADE
 ,CONSTRAINT less_than_100 CHECK (id < 100)
 ,CONSTRAINT no_99 CHECK (id != 99)
);
INSERT INTO a VALUES (1), (2), (3);
INSERT INTO b VALUES (1), (2);
INSERT INTO c VALUES (1);
SET sql_safe_updates = false;
UPDATE a SET id = id*10;
SELECT name, id FROM (
  SELECT 'a' AS name, id FROM a
UNION ALL
  SELECT 'b' AS name, id FROM b
UNION ALL
  SELECT 'c' AS name, id FROM c
)
ORDER BY name, id
;
```

This resulted in one of the nodes crashing:
```
panic: interface conversion: coldata.column is [][]uint8, not []int64

goroutine 33825 [running]:
github.com/cockroachdb/cockroach/pkg/sql/exec/coldata.(*memColumn).Int64(0xc005e371c0, 0x0, 0x72d9520, 0xc005e371c0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/coldata/vec.go:210 +0x74
github.com/cockroachdb/cockroach/pkg/sql/exec/colserde.(*ArrowBatchConverter).BatchToArrow(0xc0049cc660, 0x72b7800, 0xc006617560, 0x60, 0xc00749b800, 0xc008fabda8, 0x52d4caf, 0x1000000067243a0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colserde/arrowbatchconverter.go:165 +0x864
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Outbox).sendBatches(0xc0070d8000, 0x7286760, 0xc006443410, 0xc29c1d0, 0xc004dce120, 0xc0035e0be0, 0xc0070d8000, 0xc00417bf80, 0x400c345)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/outbox.go:205 +0x143
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Outbox).runWithStream(0xc0070d8000, 0x7286760, 0xc006443410, 0xc29c1d0, 0xc004dce120, 0xc0035e0be0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/outbox.go:276 +0x104
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Outbox).Run(0xc0070d8000, 0x72866a0, 0xc005017c00, 0xc000a76630, 0x9912173700000001, 0x9ced7d919246bc64, 0xc0a372e76e, 0x0, 0xc0035e0be0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/outbox.go:147 +0x54c
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).setupVectorizedRemoteOutputStream.func1.1(0xc0070d8000, 0x72866a0, 0xc005017c00, 0xc00333fe00, 0xc004d456e0, 0xc0035e0be0, 0xc00333ff98)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/column_exec_setup.go:813 +0x8c
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).setupVectorizedRemoteOutputStream.func1
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/column_exec_setup.go:812 +0x87
```

**Comment by yuzefovich** [comment]
A similar crash
```
SELECT array_agg(col1 ORDER BY col1), array_agg(col1 ORDER BY col1) FILTER (WHERE col1 < 0) FROM tab;
panic: interface conversion: coldata.column is []bool, not []int64

goroutine 4824 [running]:
github.com/cockroachdb/cockroach/pkg/sql/exec/coldata.(*memColumn).Int64(0xc004d5a2c0, 0x0, 0x73012a0, 0xc004d5a2c0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/coldata/vec.go:210 +0x74
github.com/cockroachdb/cockroach/pkg/sql/exec/colserde.(*ArrowBatchConverter).BatchToArrow(0xc0034d1980, 0x72df320, 0xc002125bf0, 0x60, 0xc00361a800, 0xc004d83da8, 0x52da23f, 0x100000006744860)

```
occurs on the following logic test:
```
# LogicTest: fakedist-opt-vec

statement ok
CREATE TABLE tab (
  col1 INT PRIMARY KEY,
  col2 INT,
  col3 STRING
)

# Ordered aggregations when there are no rows.
query I
SELECT array_agg(col1 ORDER BY col2) FROM TAB
----
NULL

statement ok
INSERT INTO tab VALUES (-3, 7, 'a'), (-2, 6, 'a'), (-1, 5, 'a'), (0, 7, 'b'), (1, 5, 'b'), (2, 6, 'b')

query TT colnames
SELECT array_agg(col1 ORDER BY col1), array_agg(col1 ORDER BY col1) FILTER (WHERE col1 < 0) FROM tab
----
array_agg         array_agg
{-3,-2,-1,0,1,2}  {-3,-2,-1}
```
if we add `fakedist-opt-vec` config as follows:
```
{name: "fakedist-opt-vec", numNodes: 3, useFakeSpanResolver: true,
overrideDistSQLMode: "on", overrideOptimizerMode: "on",
overrideAutoStats: "false", overrideExpVectorize: "on"},
```

**Comment by yuzefovich** [comment]
I'll be taking a look at this now.

**Comment by yuzefovich** [comment]
I added some logging (and also started generated `flowID`s for local flows so that it is easier to distinguish them, but this is not relevant here). Here is the output:
```
SELECT array_agg(col1 ORDER BY col1) FILTER (WHERE col1 < 0) FROM tab;
adding vectorized outbox for flowID {[163 16 48 48 148 96 76 15 147 32 238 205 105 208 92 216]} {Type:REMOTE StreamID:0 TargetNodeID:1}
running vectorized outbox for flowID {[163 16 48 48 148 96 76 15 147 32 238 205 105 208 92 216]} {Type:REMOTE StreamID:0 TargetNodeID:1}
adding vectorized outbox for flowID {[163 16 48 48 148 96 76 15 147 32 238 205 105 208 92 216]} {Type:REMOTE StreamID:1 TargetNodeID:1}
running vectorized outbox for flowID {[163 16 48 48 148 96 76 15 147 32 238 205 105 208 92 216]} {Type:REMOTE StreamID:1 TargetNodeID:1}
cleaning startables for flowID {[163 16 48 48 148 96 76 15 147 32 238 205 105 208 92 216]}
panic: interface conversion: coldata.column is []bool, not []int64

goroutine 1759 [running]:
github.com/cockroachdb/cockroach/pkg/sql/exec/coldata.(*memColumn).Int64(0xc00152bfc0, 0x0, 0x7302be0, 0xc00152bfc0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/coldata/vec.go:210 +0x74
github.com/cockroachdb/cockroach/pkg/sql/exec/colserde.(*ArrowBatchConverter).BatchToArrow(0xc00318c3c0, 0x72e0c60, 0xc0030baba0, 0x1, 0xc0031d4800, 0xc00320dd70, 0x52da77f, 0x100000006745d40)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colserde/arrowbatchconverter.go:165 +0x864
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Outbox).sendBatches(0xc0015e6c40, 0x72afac0, 0xc0030bac30, 0xddab390, 0xc0001fde90, 0xc00312e230, 0xc0015e6c40, 0xc002bab440, 0x400c585)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/outbox.go:205 +0x143
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Outbox).runWithStream(0xc0015e6c40, 0x72afac0, 0xc0030bac30, 0xddab390, 0xc0001fde90, 0xc00312e230)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/outbox.go:276 +0x104
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Outbox).Run(0xc0015e6c40, 0x72afa00, 0xc001d46c00, 0xc000f86de0, 0x303010a300000001, 0xcdee20930f4c6094, 0xd85cd069, 0x1, 0xc00312e230)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/outbox.go:147 +0x54c
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).setupVectorizedRemoteOutputStream.func1.1(0xc001bd8d20, 0xc001bd50e0, 0xc0015e6c40, 0x72afa00, 0xc001d46c00, 0xc00312e230, 0xc001bd8ec0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/column_exec_setup.go:865 +0x185
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).setupVectorizedRemoteOutputStream.func1
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/column_exec_setup.go:863 +0x87
```
Here is my understanding:
1. the query has a window function (although it is an aggregate here, but because of `ORDER BY`, we plan a window function) which we do not support.
2. we first plan `colBatchScan`s
3. then setup a vectorized input synchronizer (this step is "adding vectorized outbox ...")
4. we encounter a window function, so we fail the vectorized flow
5. we do `f.startables = nil` (this step is "cleaning startables ...")
However, for some reason the vectorized outboxes for the input synchronizer from step 3 are still being run, and they crash.

**Comment by yuzefovich** [comment]
Here is a very confusing to me observation:
```
SELECT array_agg(col1 ORDER BY col1) FILTER (WHERE col1 < 0) FROM tab;
flow 0xc0015cd2c0	ADDING vectorized outbox for flowID {[71 251 158 96 80 219 72 103 164 131 46 35 56 111 22 99]} {Type:REMOTE StreamID:0 TargetNodeID:1}
flow 0xc0015cd2c0	RUNNING vectorized outbox for flowID {[71 251 158 96 80 219 72 103 164 131 46 35 56 111 22 99]} {Type:REMOTE StreamID:0 TargetNodeID:1}
flow 0xc001f645a0	vectorization failed
adding ROW outbox for flowID {[222 127 23 243 131 188 77 41 135 106 63 169 40 146 114 191]}
panic: interface conversion: coldata.column is []bool, not []int64
```
I'm printing out the address where `f *Flow` is pointing at, and the addresses are different for the vectorized outbox and for the flow on which we do `f.startables = nil` (in fact, it's always `nil` at that point).

**Comment by yuzefovich** [comment]
I believe this to be fixed #38983, so I'm closing this.
